Issue 20927 fix resolves read_sas error for dates/datetimes greater than 2262-04-11

paul-lilley opened the pull request:
- [x] closes #20927 
- [x] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x ] whatsnew entry

paullilley committed:
BUG GH20927 Fixes read_sas error for dates/datetimes greater than 2262-04-11

paullilley committed:
BUG GH20927 Fixes read_sas error for dates/datetimes greater than 2262-04-11

paullilley committed:
DOC fixed typo

paul-lilley committed:
Merge branch 'master' into issue_20927_fix

mroeschke commented:
Couple of thoughts:

1) It's more idiomatic for pandas use Periods to work with out of range dates, so we may want to return period objects in this case: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#representing-out-of-bounds-spans

2) We should probably raise a `UserWarning` nonetheless to warn the user that different data dtype is being returned in this case.

paullilley committed:
fixed failing test due to column ordering

paullilley committed:
black style fix

paul-lilley commented:
Thanks for the thoughts @mroeschke 

I'll add the UserWarning about returning different types.  

As the code is at the moment, it will return datetime.date/datetime for _all_ rows if there is at least one SAS date/time that is > 2262-04-11.  
However, if read_sas is called as an iterator, chunks with all SAS date/time < 2262-04-11 will return with np.datetime64 and other chunks with at least one SAS date/time that is > 2262-04-11 will return with datetime.date/datetime - i.e. the caller will get a mix of datatypes.

I get some odd results for Periods with dates larger than 2262-04-11, e.g.
``` python
import pandas as pd
g = pd.Period(year=2262, month=4, day=11, freq='D')
print(g.start_time)
g2 = pd.Period(year=2262, month=4, day=12, freq='D')
print(g2.start_time)
```
gives
2262-04-11 00:00:00 which is correct, but then
1677-09-21 00:25:26.290448384 which is clearly wrong for 2262-04-12

This is maybe a bug in Period (?) but for the time being returning datetime.date/time gives a better representation of the original SAS dataset

mroeschke commented:
Could you open up a new issue for the Period overflowing? That looks like a bug.

`datetime.date` do not have a lot of support in pandas compared to `datetime.datetime`, so I am a little weary of returning those objects.

paul-lilley commented:
Ok, I'll raise a new issue for the Period overflow.

I'll also alter read_sas to return datetime.datetime for both dates and datetimes > 2262-04-11.  There will still be the possibility of a mix of np.datetime64 and datetime.datetime types in the final dataset if read_sas is called as an iterator.

paullilley committed:
Return datetime.datetime and raise UserWarning for all dates/datetimes > pd.Timestamp.max
Also added test for iterator behaviour,
chunks that have all dates/datetimes < pd.Timestamp.max will return np.datetime64 as usual,
chunks with any date/datetime > pd.Timestamp.max will return datetime.datetime

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

Merged upstream master

pep8speaks commented:
Hello @paul-lilley! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/io/sas/sas7bdat.py`](https://github.com/pandas-dev/pandas/blob/6b3c175434489b723a228133b25a2611134f5d82/pandas/io/sas/sas7bdat.py):

> [Line 708:89](https://github.com/pandas-dev/pandas/blob/6b3c175434489b723a228133b25a2611134f5d82/pandas/io/sas/sas7bdat.py#L708): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 88 characters)
> [Line 716:89](https://github.com/pandas-dev/pandas/blob/6b3c175434489b723a228133b25a2611134f5d82/pandas/io/sas/sas7bdat.py#L716): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 88 characters)

paullilley committed:
fixed PEP8 formating issue

TomAugspurger commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -72,6 +72,7 @@ I/O
 
 - Avoid calling ``S3File.s3`` when reading parquet, as this was removed in s3fs version 0.3.0 (:issue:`27756`)
 - Better error message when a negative header is passed in :func:`pandas.read_csv` (:issue:`27779`)
+- :func:`read_sas()` now handles dates and datetimes larger than pd.Timestamp.max returning them as datetime.datetime objects (:issue:`20927`)
```
pd.Timestamp.max should be something like :attr:`\Timestamp.max\`.

datetime.datetime should be :class:\`date time.datetime`.

paul-lilley replied:
Hi Tom,

Thanks for the comment, I'll fix the documentation tags.

mroeschke mentioned above that out of bounds datetimes are often handled as Periods
[https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#representing-out-of-bounds-spans](url)
However there is an issue with Periods [https://github.com/pandas-dev/pandas/issues/28104](url) which prevents me using them at the moment.

TomAugspurger reviewed (commented):
What do our other parsers do for out of bounds datetimes?

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
DOC fixed attribute and class markdown

paullilley committed:
fixed linting issue

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

TomAugspurger commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -703,15 +705,38 @@ def _chunk_to_dataframe(self):
+                    )
                     if self.column_formats[j] in const.sas_date_formats:
-                        unit = "d"
+                        try:
```
These try / excepts are likely complicated enough to warrant a refactor into a standalone function.

TomAugspurger commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -703,15 +705,38 @@ def _chunk_to_dataframe(self):
                 rslt[name] = np.asarray(rslt[name], dtype=np.float64)
                 if self.convert_dates:
-                    unit = None
+                    warn_msg = (
```
Does the user have some way to avoid this warning? Ideally, we would support all ov

1. warning when falling back to datetime.datetime (default)
2. Using datetime.datetime by default
3. Raising when we can't convert to datetime64

paul-lilley replied:
Hi @TomAugspurger,
I can do this but I think it would require an extra parameter to the read_sas function to specify the date/datetime handling.  In a previous approach to fixing this issue I simply surfaced the underlying boolean convert_dates parameter from SAS7BDATReader up to pandas.read_sas, but @jreback preferred not to extend the number of parameters to read_sas.
I see 2 ways forward: 
1) add a new parameter to specify the date_handling format,
2) drop the warning and just return datetime64 when we can and datetime.datetime when we exceed the bounds of datetime64 (which is reasonably rare)

Which would you and Jeff prefer?

TomAugspurger replied:
Any thoughts on this @jreback? I slightly prefer an extra parameter to control the behavior of out-of-bounds timestamps, but could go either way.

TomAugspurger commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -72,6 +72,7 @@ I/O
 
 - Avoid calling ``S3File.s3`` when reading parquet, as this was removed in s3fs version 0.3.0 (:issue:`27756`)
 - Better error message when a negative header is passed in :func:`pandas.read_csv` (:issue:`27779`)
+- :func:`read_sas()` now handles dates and datetimes larger than :attr:`Timestamp.max` returning them as :class:`datetime.datetime` objects (:issue:`20927`)
```
This will need to be moved to 1.0.0.rst

paul-lilley replied:
Moved

jreback commented:
can you merge master and update to comments

paul-lilley commented:
Hi, apologies for the delayed response.  I'll try to address Tom's points, then merge master and update the comments.

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
CLN: minor refactor for reading sas files, updated whatsnew

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

TomAugspurger reviewed (commented):
Looks like there are some linting errors @paul-lilley: https://github.com/pandas-dev/pandas/pull/28047/checks?check_run_id=275107697

jreback commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -687,6 +689,39 @@ def _read_next_page(self):
+            # SAS float64 lacks precision for more than ms resolution so the
+            # fit to datetime.datetime is ok
+            warn(
+                "SAS datetime > pandas.Timestamp.max, returning "
```
is there a reason to have 2 very similar routines here?

paul-lilley replied:
Hi,  I've combined the two into a single function

jreback reviewed (changes requested):
our default behavior of to_datetime is to convert to datetime when OOB
so i think it would be ok to just make this change w/o a keyword arg

paullilley committed:
removed warnings

paullilley committed:
merged upstream/master

paullilley committed:
tidied trailing whitespace

jreback commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -687,6 +687,26 @@ def _read_next_page(self):
         return False
 
+    @staticmethod
+    def _convert_datetimes(sas_datetimes, unit):
```
can you type the args format the doc-string like others

jreback commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -687,6 +687,26 @@ def _read_next_page(self):
 
         return False
 
+    @staticmethod
```
rather than a static method can you just make this module level

paul-lilley replied:
OK

jreback commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -687,6 +687,26 @@ def _read_next_page(self):
 
+    @staticmethod
+    def _convert_datetimes(sas_datetimes, unit):
+        """Converts to np.datetime64 if possible, otherwise to dateime.datetime
```
we *never* convert to np.datetime64, no need to mention this, rather we convert to Timestamp (misspell on datetime)

paul-lilley replied:
I've updated the docstring

jreback commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -687,6 +687,26 @@ def _read_next_page(self):
+        to datetime.datetime is ok
+        """
+        try:
+            return pd.to_datetime(sas_datetimes, unit=unit, origin="1960-01-01")
```
we *already* have an ``errors`` keyword can you use this?

paul-lilley replied:
Hi,

Not quite sure what you mean - are you thinking that the call to to_datetime (line 698 above) should have errors='ignore' and then use the apply to pick up any items that are still float64 (from SAS) and convert just those items to datetime.datetime?

I went for the try/except as the majority of SAS datasets probably only have dates within the bounds of Timestamp, and we then avoid the additional cost of the apply to check for remaining float64 values to convert.

On a slight aside, I notice also that the docstring for pd.to_datetime states
"In case when it is not possible to return designated types (e.g. when
        any element of input is before Timestamp.min or after Timestamp.max)
        return will have datetime.datetime type (or corresponding
        array/Series)."
This doesn't seem to be true, e.g.
import pandas as pd
import datetime

df = pd.DataFrame([{'max': '9999-12-31'}])
df['conv'] = pd.to_datetime(df['max'])

print(df)

raises OutOfBoundsDatetime rather than converting to a datetime.datetime for 9999-12-31

paul-lilley replied:
Hi @jreback, I've merged master - could you take a look? Thanks.

jreback commented:
can you merge master and will look again

WillAyd commented:
Closing as I think stale, but @paul-lilley ping if you can address comments and would like to pick back up

paullilley committed:
moved static method to module level function

paullilley committed:
merged master

paul-lilley commented:
Hi @WillAyd 
I think this is still an issue - I've updated the code and merged master, can you reopen the request?
Thanks

paullilley committed:
fixed import sorting

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -1186,6 +1186,7 @@ I/O
 - Bug in :func:`pandas.io.json.json_normalize` where a missing value in the location specified by `record_path` would raise a ``TypeError`` (:issue:`30148`)
 - :func:`read_excel` now accepts binary data (:issue:`15914`)
 - Bug in :meth:`read_csv` in which encoding handling was limited to just the string `utf-16` for the C engine (:issue:`24130`)
+- :func:`read_sas()` now handles dates and datetimes larger than :attr:`Timestamp.max` returning them as :class:`datetime.datetime` objects (:issue:`20927`)
```
can you move to 1.1

paul-lilley replied:
Done.

jreback commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -214,3 +214,83 @@ def test_zero_variables(datapath):
+    #    sas7bdat module
+    fname = datapath("io", "sas", "data", "max_sas_date.sas7bdat")
+    df = pd.read_sas(fname, encoding="iso-8859-1")
+    # SAS likes to left pad strings with spaces - lstrip before comparing
```
can you put a blank line before the comment

paul-lilley replied:
Done.

jreback reviewed (changes requested):
looks good. ping on green. (merge master as well)

jreback commented:
@bashtage if any comments.

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
moved whatsnew to 1.1, added space before comment

bashtage commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -214,3 +214,84 @@ def test_zero_variables(datapath):
+                "text": ["normal"],
+                "dt_as_float": [1880323199.999],
+                "dt_as_dt": [np.datetime64("2019-08-01 23:59:59.999")],
+                "date_as_float": [float(21762)],
```
Same comment

paul-lilley replied:
Done

bashtage commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -214,3 +214,84 @@ def test_zero_variables(datapath):
+                datetime(9999, 12, 29, 23, 59, 59, 999000),
+                datetime(2019, 8, 1, 23, 59, 59, 999000),
+            ],
+            "date_as_float": [float(2936547), float(21762)],
```
Same comment.

paul-lilley replied:
Done

bashtage reviewed (approved):
Small issue, mostly a matter of taste.  Overall LGTM.

paul-lilley commented on pandas/tests/io/sas/test_sas7bdat.py:
```diff
@@ -214,3 +214,84 @@ def test_zero_variables(datapath):
+                "text": ["max"],
+                "dt_as_float": [253717747199.999],
+                "dt_as_dt": [datetime(9999, 12, 29, 23, 59, 59, 999000)],
+                "date_as_float": [float(2936547)],
```
Hi @bashtage,
SAS has only float and fixed length string datatypes, the explicit cast was to highlight this.
Howerver, I see your point as the dict key also highlights that it is a float.
I'll tidy it up.

paul-lilley replied:
Done

paullilley committed:
tidied float constants

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
fixed linting

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paul-lilley commented:
Hi @jreback 
I've made the minor changes from @bashtage and merged master, can you take a look? Thanks.

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

WillAyd commented:
Looks like this one fell through but is close - @paul-lilley can you fix the merge conflict?

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paul-lilley commented:
Hi @WillAyd - I've merged master (it went through without merge conflicts)

bashtage commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -29,6 +29,31 @@
+    ----------
+    sas_datetimes : Series of 64bit floats representing dates or datetimes
+        in SAS
+    unit : "d" if the floats represent dates, "t" for datetimes
```
Description does not match the use below.  unit is either s or d.  Typo?

Also, these are not correctly formatted.  Should be more liek

```
sas_datetime : {Series, Sequence[float]}
   Dates or datetimes in SAS
....
```

Of course this is internal, but might as well be consistent when possible.

paul-lilley replied:
Good spot - I've altered the comment

bashtage reviewed (changes requested):
Noticed that the docstring does not quite align with the code.

paullilley committed:
fixed failing test

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
fixed failing test

paullilley committed:
fixed linting

paullilley committed:
fixed import order

paul-lilley commented:
Hi @WillAyd , @bashtage I've altered the comment and merged master

jreback commented on pandas/io/sas/sas7bdat.py:
```diff
@@ -29,6 +29,32 @@
 
 
+def _convert_datetimes(sas_datetimes: pd.Series, unit: str):
+    """
```
can you add a return type

paul-lilley replied:
Done.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -773,6 +773,7 @@ I/O
   timestamps with ``version="2.0"`` (:issue:`31652`).
 - Bug in :meth:`read_csv` was raising `TypeError` when `sep=None` was used in combination with `comment` keyword (:issue:`31396`)
 - Bug in :class:`HDFStore` that caused it to set to ``int64`` the dtype of a ``datetime64`` column when reading a DataFrame in Python 3 from fixed format written in Python 2 (:issue:`31750`)
+- :func:`read_sas()` now handles dates and datetimes larger than :attr:`Timestamp.max` returning them as :class:`datetime.datetime` objects (:issue:     `20927`)
```
can you remove this extra space before the issue number

paul-lilley replied:
Done

jreback reviewed (approved):
2 minor comments, ping on green

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

paullilley committed:
Added return type. Removed space before issue number.

paullilley committed:
Fixed mypy error.

paullilley committed:
Merge remote-tracking branch 'upstream/master' into issue_20927_fix

jreback commented:
thanks @paul-lilley very nice!

jbrockmendel commented:
@paul-lilley im trying to understand the comment on test_max_sas_date

```
    # NB. max datetime in SAS dataset is 31DEC9999:23:59:59.999
    #    but this is read as 29DEC9999:23:59:59.998993 by a buggy
    #    sas7bdat module
```

Does this mean that the test here is still wrong and it really should be coming back with `datetime(9999, 12, 31, 23, 59, 59, 999000)`?

jbrockmendel commented:
@bashtage do you have convenient access to SAS and willingness to help me track down https://github.com/pandas-dev/pandas/pull/28047#issuecomment-1780391974

paullilley commented:
Hi @jbrockmendel
Correct, it really should be coming back with datetime(9999, 12, 31, 23, 59, 59, 999000) as max datetime in SAS dataset is 31DEC9999:23:59:59.999 (ie. millisecs precision not microsecs).
However, I'd recommend moving all the SAS dataset reading to use https://github.com/Roche/pyreadstat (currently used in Pandas for SPSS data) as it is way faster

bashtage commented:
I don't have access to SAS easily, but @paullilley answered.

paullilley commented:
I do have access to SAS if needed.

jbrockmendel commented:
Thanks @paul-lilley.  When I read the file with pyreadstat it comes back as `datetime(9999, 12, 29, 23, 59, 59)`, i.e. the day is 29 instead of 31 and the seconds are rounded down.  Is there a chance the file was written with Dec 29 instead of Dec 31, or is there a problem that common to both the pandas and pyreadstat readers?
